File: placement-reactnative/storyly-placement-react-native/ios/Common/Data/PlacementConfigData.swift
import Foundation
import StorylyPlacement
import StorylyCore

func decodeSTRPlacementConfig(_ config: [String: Any], token: String) -> STRPlacementConfig {
  let builder = STRPlacementConfig.Builder()
  
  if let testMode = config["testMode"] as? Bool {
    _ = builder.setTestMode(isTest: testMode)
  }
  
  if let locale = config["locale"] as? String {
    _ = builder.setLocale(locale: locale)
  }
  
  if let layoutDirection = config["layoutDirection"] as? String {
    _ = builder.setLayoutDirection(direction: layoutDirection == "rtl" ? .RTL : .LTR)
  }
  
  if let customParameter = config["customParameter"] as? String {
    _ = builder.setCustomParameter(parameter: customParameter)
  }
  
  if let labels = config["labels"] as? [String] {
    _ = builder.setLabels(labels: Set(labels))
  }
  
  if let userProperties = config["userProperties"] as? [String: String] {
    _ = builder.setUserProperties(data: userProperties)
  }
  
  if let productConfigDict = config["productConfig"] as? [String: Any],
     let productConfig = decodeSTRProductConfig(productConfigDict) {
    _ = builder.setProductConfig(config: productConfig)
  }
  
  if let shareConfigDict = config["shareConfig"] as? [String: Any],
     let shareConfig = decodeSTRShareConfig(shareConfigDict) {
    _ = builder.setShareConfig(config: shareConfig)
  }
  
  if let networkConfigDict = config["networkConfig"] as? [String: Any],
     let networkConfig = decodeSTRNetworkConfig(networkConfigDict) {
    _ = builder.setNetworkConfig(config: networkConfig)
  }
  
  return builder.build(token: token)
}

func decodeSTRProductConfig(_ productConfig: [String: Any]) -> STRProductConfig? {
  let builder = STRProductConfig.Builder()
  return builder.build()
}

func decodeSTRShareConfig(_ shareConfig: [String: Any]) -> STRShareConfig? {
  let builder = STRShareConfig.Builder()
  
  if let facebookAppId = shareConfig["facebookAppId"] as? String {
    _ = builder.setFacebookAppID(id: facebookAppId)
  }
  
  if let shareUrl = shareConfig["shareUrl"] as? String {
    _ = builder.setShareUrl(url: shareUrl)
  }
  
  if let appLogoVisibility = shareConfig["appLogoVisibility"] as? Bool {
    _ = builder.setAppLogoVisibility(isVisible: appLogoVisibility)
  }
  
  return builder.build()
}

func decodeSTRNetworkConfig(_ networkConfig: [String: Any]) -> STRNetworkConfig? {
  let builder = STRNetworkConfig.Builder()
  
  if let cdnHost = networkConfig["cdnHost"] as? String {
    _ = builder.setCdnHost(url: cdnHost)
  }
  
  if let productHost = networkConfig["productHost"] as? String {
    _ = builder.setProductHost(url: productHost)
  }
  
  if let analyticHost = networkConfig["analyticHost"] as? String {
    _ = builder.setAnalyticHost(url: analyticHost)
  }
  
  if let placementHost = networkConfig["placementHost"] as? String {
    _ = builder.setListHost(url: placementHost)
  }
  
  return builder.build()
}

File: placement-reactnative/storyly-placement-react-native/android/src/main/java/com/storylyplacementreactnative/common/data/PlacementConfigData.kt
package com.storylyplacementreactnative.common.data

import com.appsamurai.storyly.core.config.placement.STRLayoutDirection
import com.appsamurai.storyly.core.config.placement.STRNetworkConfig
import com.appsamurai.storyly.core.config.placement.STRPlacementConfig
import com.appsamurai.storyly.core.config.placement.STRProductConfig
import com.appsamurai.storyly.core.config.placement.STRShareConfig
import com.storylyplacementreactnative.common.data.product.decodeSTRProductItem


fun decodeSTRPlacementConfig(config: Map<String, Any?>, token: String): STRPlacementConfig {
    val builder = STRPlacementConfig.Builder()

    (config["testMode"] as? Boolean)?.let { builder.setTestMode(it) }
    (config["locale"] as? String)?.let { builder.setLocale(it) }
    (config["layoutDirection"] as? String)?.let { layoutDir ->
        builder.setLayoutDirection(
            if (layoutDir == "rtl") STRLayoutDirection.RTL else STRLayoutDirection.LTR
        )
    }
    (config["customParameter"] as? String)?.let { builder.setCustomParameter(it) }
    (config["labels"] as? List<*>)?.filterIsInstance<String>()?.let { labels ->
        builder.setLabels(labels.toSet())
    }
    (config["userProperties"] as? Map<String, String>)?.let { userProps ->
        builder.setUserProperties(userProps)
    }
    decodeSTRProductConfig(config["productConfig"] as? Map<String, Any?>)?.let { productConfig ->
        builder.setProductConfig(productConfig)
    }
    decodeSTRShareConfig(config["shareConfig"] as? Map<String, Any?>)?.let { shareConfig ->
        builder.setShareConfig(shareConfig)
    }
    decodeSTRNetworkConfig(config["networkConfig"] as? Map<String, Any?>)?.let { networkConfig ->
        builder.setNetworkConfig(networkConfig)
    }
    return builder.build(token)
}

fun decodeSTRProductConfig(productConfig: Map<String, Any?>?): STRProductConfig? {
    productConfig ?: return null
    val builder = STRProductConfig.Builder()
    return builder.build()
}

fun decodeSTRShareConfig(shareConfig: Map<String, Any?>?): STRShareConfig? {
    shareConfig ?: return null
    val config = STRShareConfig.Builder()
    (shareConfig["facebookAppId"] as? String)?.let { config.setFacebookAppID(it) }
    (shareConfig["shareUrl"] as? String)?.let { config.setShareUrl(it) }
    (shareConfig["appLogoVisibility"] as? Boolean)?.let { config.setAppLogoVisibility(it) }
    return config.build()
}

fun decodeSTRNetworkConfig(networkConfig: Map<String, Any?>?): STRNetworkConfig? {
    networkConfig ?: return null
    val config = STRNetworkConfig.Builder()
    (networkConfig["cdnHost"] as? String)?.let { config.setCdnHost(it) }
    (networkConfig["productHost"] as? String)?.let { config.setProductHost(it) }
    (networkConfig["analyticHost"] as? String)?.let { config.setAnalyticHost(it) }
    (networkConfig["placementHost"] as? String)?.let { config.setPlacementHost(it) }
    return config.build()
}